spec/FrameSpec.js

```diff
@@ -0,0 +1,299 @@
+    frame.play(3);
+    frame.play(7);
+    expect(frame._bonusBalls).toEqual(1);
+    expect(frame._score).toEqual(10);
+    expect(frame.isComplete()).toEqual(true);
+  });
+
+  it("shouldn't set bonus balls if not a strike or spare 3&6", function() {
+    frame.play(3);
+    frame.play(6);
+    expect(frame._bonusBalls).toEqual(0);
+    expect(frame._score).toEqual(9);
+    expect(frame.isComplete()).toEqual(true);
+  });
+
+
+  it("should set balls remaining to two on last frame if a strike", function() {
+    lastFrame.play(10);
+    expect(lastFrame._ballsRemaining).toEqual(2);
+  });
+
+  it("should set balls remaining to one on last frame if a spare", function() {
+    lastFrame.play(3);
+    lastFrame.play(7);
+    expect(lastFrame._ballsRemaining).toEqual(1);
+  });
+
+  describe("Last frame not a strike or spare", function() {
+
+    beforeEach(function() {
+      lastFrame = new Frame(true);
+      lastFrame.play(3);
+      lastFrame.play(6);
+    });
+
+    it("shouldn't set bonus balls on the last frame if not a strike or spare", function() {
+      expect(lastFrame._bonusBalls).toEqual(0);
+    });
+
+    it("shouldn't set balls remaining on the last frame if not a strike or spare", function() {
+      expect(lastFrame._ballsRemaining).toEqual(0);
+    });
+
+    it("should set score to 9 on last frame with 3 & 6 pins", function() {
+      expect(lastFrame._score).toEqual(9);
+    });
+
+    it("should set frame complete after two balls on last frame if score <10", function() {
+      expect(lastFrame.isComplete()).toEqual(true);
+    });
+
+  });
+  describe("Ensure correct pins added for each ball", function() {
+
+    beforeEach(function() {
+      frame = new Frame();
+    });
+
+    it("should set ball1 to X after strike", function() {
+      frame.play("X");
+      expect(frame.ball1).toEqual(10);
+    });
+    it("should set ball1 to - after first play", function() {
+      frame.play("-");
+      expect(frame.ball1).toEqual(0);
+    });
+    it("should set ball1 to 7 after first play", function() {
+      frame.play(7);
+      expect(frame.ball1).toEqual(7);
+    });
+    it("should set ball2 to 10 after spare", function() {
+      frame.play("-");
+      frame.play("/");
+      expect(frame.ball2).toEqual(10);
+    });
+    it("should set ball2 to 7 after spare 3&7", function() {
+      frame.play(3);
+      frame.play(7);
+      expect(frame.ball2).toEqual(7);
+    });
+    it("should set ball2 to 6 after second play", function() {
+      frame.play(3);
+      frame.play(6);
+      expect(frame.ball2).toEqual(6);
+    });
+  });
+
+  describe("Pins remaining", function() {
+
+    beforeEach(function() {
+      frame = new Frame();
+    });
+
+    it("should be 10 before play starts", function() {
+      expect(frame.pinsRemaining()).toEqual(10);
+    });
+
+    it("should be 5 if 5 already down", function() {
+      frame.play(5);
+      expect(frame.pinsRemaining()).toEqual(5);
```

Comment:
Hmm. So I have written and rewritten this comment a couple of times. I feel like there are a lot of tests here and some of them are duplicating but its hard for me to pin down exactly what I would change. 

I feel like these three tests are a good example - I think there's a single test here (if I play 4 and then 3, there are 3 pins remaining, for example) which basically covers all of the cases - at least, I can't imagine many regressions I could introduce which would cause any of these tests to fail and not that one.

There are a couple of other examples below which I will comment. Testing is good, but too much testing makes code brittle - if I have multiple tests testing a single flow through my application, if I want to make a change to the application then I have to fix multiple tests when I refactor. I want my tests to help identify regressions, but I don't want them to hinder me from refactors, so there is a balance there.